File: 350. Intersection of Two Arrays II.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> map;
        vector<int> result;
        for(auto val : nums1) {
            map[val] ++;
        }
        for(auto val : nums2) {
            if(map.count(val) > 0 && map[val] > 0) {
                map[val] --;
                result.push_back(val);
            }
        }
        return result;
    }
};

File: 350. Intersection of Two Arrays II.java
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        HashMap<Integer, Integer> map = new HashMap<>();
        ArrayList<Integer> intersect = new ArrayList<Integer>();

        for(int val : nums1) {
            if(!map.containsKey(val))map.put(val, 1);
            else map.put(val, map.get(val) + 1);
        }

        for(int val : nums2) {
            if(map.containsKey(val) && map.get(val) > 0) {
                intersect.add(val);
                map.put(val, map.get(val) - 1);
            }
        }

        int[] result = new int[intersect.size()];
        int i = 0;
        for(int val : intersect) {
            result[i ++] = val;
        }

        return result;
    }
}
